Hcsvlab 1096

shaunmuscat opened the pull request.

shaunmuscat committed:
HCSVLAB-1096: add interface to filter item list document download

Update the web interface with the facility to filter the documents that are downloaded when downloading an item list.

shaunmuscat committed:
HCSVLAB-1087: update API to filter files downloaded

Update the API for downloading items so that it can take a regular expression parameter to filter the documents that are downloaded.

shaunmuscat committed:
HCSVLAB-1096: add item list filter for aspera download

shaunmuscat committed:
HCSVLAB-1096: update download as zip

dyazbek-intersect commented on lib/item/download_items_helper.rb:
```diff
@@ -480,6 +483,17 @@ def generate_json_log(valids, invalids)
       json_log.to_json
     end
 
+    def filter_item_files(filenames, regexp_filter=DEFAULT_DOCUMENT_FILTER)
```
This method needs some heavy unit testing.

shaun-intersect replied:
What kind of unit testing would be needed? It would be expected that a valid Regexp object is being passed to the function, in addition to a syntactically correct array of filename strings. This method then just returns whichever of those files whose basename has some match against the Regexp.
Indirect testing with actual (valid and invalid) regular expressions happens in features/api.feature:638-675.

dyazbek-intersect replied:
Given a long array of various filenames that you have created in your test I'd be checking what comes out is what is expected.

e.g.:
filenames=['cat.cat','foo','']
asserting: *cat *.cat cat* cat.* food* catcat* etc. etc. etc.
Much easier to test at the method level than it would be at the controller level.

dyazbek-intersect commented on features/download_item_list.feature:
```diff
@@ -0,0 +1,49 @@
+    And I should see "txt"
+    And I should see "Download Selected"
+    And I should see "Download only files that match a particular regular expression"
+    And I should see "Download Matches"
```
Would be a good idea to add a test that checks expected content of the ZIP file to ensure all three types of filters are working.

shaun-intersect replied:
The ZIP file couldn't be obtained using the feature tests as it couldn't find/click the 'download all/selected/matches' buttons and the API returns the bytes of the ZIP rather than storing it in some location.

dyazbek-intersect commented:
The JIRA clearly asks you to implement Ruby regular expressions as input - and you have done a great job at this, but I would be asking weather or not this is what the user really wants...
Typically when selecting files, users are more familiar with the GLOB style (https://en.wikipedia.org/wiki/Glob_%28programming%29)...

If Ruby regular expressions are really what they want then I would suggest putting examples into the dialog and even linking to a documentation on how to form ruby regular expressions

shaun-intersect commented:
Pushed the code for https://jira.intersect.org.au/browse/HCSVLAB-1087 which is the corresponding API calls for the web interface in this story. Tests with expected actual regular expressions are included.

shaunmuscat committed:
HCSVLAB-1096: update example Regexp in web interface

dyazbek-intersect commented on features/api.feature:
```diff
@@ -630,11 +630,50 @@ Feature: Browsing via API
+    Then I should get a 400 response code
+    And the JSON response should be:
+    """
+    {"error":"Error with regular expression: target of repeat operator is not specified: /*.txt/"}
```
Why is *.txt a 400?

shaun-intersect replied:
It isn't a legitimate regex, since the * is a repeater not a wildcard (and no symbol is given before it to match repetitions of) and 400 seemed the most appropriate error code to use. 
I'm not sure if *.ext is a valid glob however.

dyazbek-intersect replied:
You are right - but regex and glob are used interchangeably.
I honestly think that if you demonstrate this story in it's current state to the client then it's going to come back for rework because most people think a file-selecting regex should look like *.txt rather than .*\.txt$

dyazbek-intersect commented:
http://ruby-doc.org/core-2.2.3/File.html#method-c-fnmatch Use this method to do globbing - regular expressions aren't what we need for this task.

shaunmuscat committed:
HCSVLAB-1096: update filter to use glob

Update file filter to use glob rather than regex, and add associated rspec tests

dyazbek-intersect commented:
Good job :+1: 
Thanks for being patient
